mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostClassificationModelParitySpec.scala

```diff
@@ -0,0 +1,128 @@
+import ml.dmlc.xgboost4j.scala.Booster
+import org.apache.spark.ml.linalg.Vectors
+import org.scalatest.FunSpec
+
+
+class XGBoostClassificationModelParitySpec extends FunSpec
+  with BoosterUtils
+  with CachedDatasetUtils
+  with BundleSerializationUtils
+  with FloatingPointApproximations {
+
+  def createBoosterClassifier(booster: Booster): Transformer ={
+
+    XGBoostClassification(
+      "xgboostSingleThread",
+      NodeShape.probabilisticClassifier(
+        rawPredictionCol = Some("raw_prediction"),
+        probabilityCol = Some("probability")),
+      XGBoostClassificationModel(XGBoostBinaryClassificationModel(booster, numFeatures, 0))
+    )
+  }
+
+  def createMultinomialBoosterClassifier(booster: Booster): Transformer ={
+
+    XGBoostClassification(
+      "xgboostSingleThread",
+      NodeShape.probabilisticClassifier(
+        rawPredictionCol = Some("raw_prediction"),
+        probabilityCol = Some("probability")),
+      XGBoostClassificationModel(XGBoostMultinomialClassificationModel(booster, 3, numFeatures, 0))
+    )
+  }
+
+  def equalityTestRowByRow(booster: Booster, mleapTransformer: Transformer) = {
+
+    import XgbConverters._
+
+    leapFrameLibSVMtest.dataset.foreach {
+      r=>
+        val mleapResult = mleapTransformer.transform(DefaultLeapFrame(mleapSchema.get, Seq(r))).get
+
+        val mleapPredictionColIndex = mleapResult.schema.indexOf("prediction").get
+        val mleapRawPredictionColIndex = mleapResult.schema.indexOf("raw_prediction").get
+        val mleapProbabilityColIndex = mleapResult.schema.indexOf("probability").get
+
+        val singleRowDMatrix = r(1).asInstanceOf[SparseTensor[Double]].asXGB
+
+        val boosterResult = booster.predict(singleRowDMatrix, false, 0).head(0)
+
+        val boosterProbability = Vectors.dense(1 - boosterResult, boosterResult).toDense
+        val boosterPrediction = Math.round(boosterResult)
+
+
+        assert (boosterPrediction == mleapResult.dataset.head.getDouble(mleapPredictionColIndex))
+
+        assert (
+          almostEqualSequences(
+            Seq(boosterProbability.values),
+            Seq(mleapResult.dataset.head.getTensor[Double](mleapProbabilityColIndex).toArray)
+          )
+        )
+
+        val boosterResultWithMargin = booster.predict(singleRowDMatrix, true, 0).head(0)
+        val boosterRawPrediction = Vectors.dense(- boosterResultWithMargin, boosterResultWithMargin).toDense
+
+        assert (almostEqualSequences(
+          Seq(boosterRawPrediction.values),
+          Seq(mleapResult.dataset.head.getTensor[Double](mleapRawPredictionColIndex).toArray)
+        ))
+    }
+  }
+
+  it("Results between the XGBoost4j booster and the MLeap Transformer are the same") {
+    val booster = trainBooster(xgboostParams, denseDataset)
+    val xgboostTransformer = createBoosterClassifier(trainBooster(xgboostParams, denseDataset))
```

Comment:
So the transformer is created via another `trainBooster` call? How do we guarantee that it ends up building the same tree? Are the xgboostParams seeded?

Reply:
Originally I was re-using the booster object but then I thought that it must probably create the same tree by construction. It seems that it does - otherwise this test would/will fail